Doc 8536 agent scope - updated travel-sample app for nodeJS

osfameron opened the pull request.

osfameron committed:
DOC-8536 start update of travel-sample to latest spec

Based on DOC-8136 (python) and DOC-8135 (frontend).

This commit:

 * removes static directory (as we'll run frontend in docker/from its own project)
 * updates deprecated bodyParser
 * updates endpoints
 * puts tenanted APIs in own Router.
   (untested, not fully plumbed in as yet)
 * adds a summary of endpoints at `/`
 * plus looks like vscode has done driveby whitespace fixes..

osfameron committed:
DOC-8536 use agent scope

Brings (mostly) inline with Python app and spec.

Note that the `hotels` search inflates the results with a KV get
with `project` parameter, rather than the subdoc manipulations
specified. But the `project` version does seem to do very similar,
so will check with PM and engineering if this is equivalent.

osfameron committed:
Dockerise, add logging

osfameron committed:
DOC-8536 fix error handling

 * move error handler to end (after all routes and use) so that it runs
 * however Async routes don't trigger handler by default.
   Use a `runAsync` wrapper to make this work.
    * NB: https://github.com/vkarpov15/awaitjs-express/ might be a good
      solution for more worked example
    * indenting all async handlers means lots of *whitespace* changes,
      view diff with `-w`
 * remove vestigial try/catch, and 500 handlers, where the global handler
   is now being used.
 * improve output
    * add comment for the INVALID_PACKET error (mostly for our benefit,
      as devs in this period where we're hitting it, but can imagine
      it will affect a few users in due course too)
 * make it clear when `res.send()` is the exit point of function, by

osfameron committed:
Semicolon Extravaganza

 * make formatting of semicolons consistent.
   (existing code used semicolons, as per https://javascript.info/coding-style
   whereas new code didn't, like https://standardjs.com/
   Don't think it matters either way, but best to be consistent.
   Have gone for no-semicolons style.

 * moved semicolons of constructed N1QL statement to where they belong
   (e.g. in the outer template)

osfameron committed:
Swagger UI

Using swagger-ui-express, and the swagger.json from the
canonical Python repository.

maria-robobug commented on Dockerfile:
```diff
@@ -0,0 +1,25 @@
+
+WORKDIR /app
+
+#RUN apt-get update && apt-get install -y \
```
Do we need this here or can we remove?

osfameron replied:
oops yes (I wasn't sure which deps this backend would need and forgot to prune...)

maria-robobug commented on index.js:
```diff
@@ -17,349 +21,408 @@ var cluster = new couchbase.Cluster(
-    return;
+  return res.send(
+    `<h1>NodeJS Travel Sample API</h1>
+    A sample API for getting started with Couchbase Server and the NodeJS SDK.
```
Looks good, will try to add in Python as well!

maria-robobug commented on index.js:
```diff
@@ -17,349 +21,408 @@ var cluster = new couchbase.Cluster(
+app.use(cors())
+app.use(express.json())
+
+app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
```
Should we have this as `/apidocs` or `/api-docs`? Just thinking it would be good to keep it consistent across projects. Currently I've got it as `/apidocs` in Python because the library I used defaults to that, but I can try to configure it if needed.

osfameron replied:
sorry, I just copy/pasted the Node library example too! I'll change to match

maria-robobug commented on index.js:
```diff
@@ -17,349 +21,408 @@ var cluster = new couchbase.Cluster(
       if (err) {
-        res.status(400).send({
+        return res.status(400).send({
           error: 'Invalid JWT token',
```
In the swagger api docs I have the error model as `{"message": "error message x"}`, is this what will be returned by the Node api? 

Or will it return something like this?
```
{"error": "Invalid JWT token", cause: "some cause"}
```

osfameron replied:
the frontend seems to only read `error` and `context` fields in the case of an error, does the python app act differently?

I'll go through the swagger tests and check the example results, haven't done that yet!

maria-robobug reviewed (commented):
Looking good! 👍

osfameron committed:
Tweaks

Addressing PR review comments.

maria-robobug commented:
Ah, I forget, we'll need to update the Readme as well. For running with Docker etc...

osfameron committed:
Script to update swagger.json

osfameron commented:
> Ah, I forget, we'll need to update the Readme as well. For running with Docker etc...

Ah yes, I'll copy the python as a template...

osfameron committed:
Copy README.md from Python backend

(with no changes, to make tweaks easier to diff)

osfameron committed:
Update README and startup

 * port README from Python to Node.js
   * plus some reordering/rewording for clarity
 * implement the CB_HOST etc. variables
   * passed *as* env vars rather than arguments
 * correct spelling of Node.js

osfameron committed:
Fix docker-compose startup ordering

Latest docker-compose doesn't support `condition`. See
e.g. https://docs.docker.com/compose/compose-file/compose-file-v3/

So we just depend on the service, which only handles service *start*
order.

The simple `/pools` endpoint that we used in the previous healthcheck
doesn't guarantee that the Couchbase server is configured properly. Not sure,
but I think it may have worked previously due to generous padding with `sleep`
(e.g. as a race condition) so have looked for something more robust.

This commit checks the `/pools/default/buckets/travel-sample/scopes`
endpoint and uses jq to ensure that the expected scopes are all
present and correct.

osfameron committed:
Tidying

osfameron committed:
Fix docker compose startup

New function `wait-for` encapsulates the logic of requesting
the API and checking data on it using `jq`.

maria-robobug commented on wait-for-couchbase.sh:
```diff
@@ -1,29 +1,106 @@
+}
+
+##### Script starts here #####
+TIMES=150
```
Not sure what `TIMES` represents, is this in seconds?

osfameron replied:
this is the number of times it attempts to poll, suggest a better name?

maria-robobug replied:
`ATTEMPTS`? Think you used it elsewhere, seems to make sense 👍

maria-robobug commented on wait-for-couchbase.sh:
```diff
@@ -1,29 +1,106 @@
+
+  for QUERY in "$@"
+  do
+    wait-for-one $ATTEMPTS $URL "$QUERY" || ( log "Failure"; exit 1001 )
```
how come exit `1001`?

osfameron replied:
just needs to be a non-zero number, but yeah, `1` is probably better...

maria-robobug commented on wait-for-couchbase.sh:
```diff
@@ -1,29 +1,106 @@
+  '. | del(.indexer) | del(.["travel-sample:def_name_type"]) | map(.items_count > 0) | all' \
+  '. | del(.indexer) | map(.num_pending_requests == 0) | all'
 
+exec $@
```
Should we not start the application after these checks?

osfameron replied:
yeah, docker-compose passes in `"node", "index.js"` now.
(Could just start it - but parametrizing it also makes it a bit easier to test locally :) )

maria-robobug commented on docker-compose.yml:
```diff
@@ -11,22 +11,24 @@ services:
       - CB_PSWD
-    entrypoint: ["./wait-for-couchbase.sh", "./entrypoint.sh"]
+      - COLUMNS
+    entrypoint: ["./wait-for-couchbase.sh", "node", "index.js"]
```
I think this script should just be called in the Dockerfile as the entrypoint no?  That way there is no need to put an entrypoint here.

osfameron replied:
ooo, could be. I haven't tested with `docker up` but should do... Will add that to todo list, ta!

maria-robobug replied:
yeah, not sure if tested already, but we want to make sure this will behave the same way with a different Couchbase Server setup (i.e it will still create the FTS index and perform checks).

For example:
 `CB_HOST=192.168.0.2 CB_USER=Administrator CB_PSWD=password docker-compose up backend frontend`

osfameron replied:
yeah, I've tested manually, to make sure it still worked as env vars instead of parameters, was one of the fun rabbitholes I went down yesterday... but definitely will re-test...

maria-robobug replied:
Yeah I just wanted to make sure it runs those checks regardless of Couchbase Server flavour, but other than that this is ace! 👍 

I'll test this out with Python and Java whooop!

osfameron replied:
Oh, so this is interesting, on e.g. `docker-compose up backend frontend`, I'm sure I've used this myself, but...

> `docker-compose up SERVICE` automatically includes `SERVICE`’s dependencies. In the example below, docker-compose up web also creates and starts `db` and `redis`.
> https://docs.docker.com/compose/compose-file/compose-file-v3/

😱 I'll do more testing, but given what's *documented* might move to advising using `docker up` manually in these cases?

osfameron replied:
huh, yeah
```
❯ docker-compose up backend
Docker Compose is now in the Docker CLI, try `docker compose up`

Starting couchbase-sandbox-7.0.0-beta ... done
Starting try-cb-api                   ... done
```

osfameron replied:
fairly sure it didn't use to do this...

maria-robobug replied:
Hmm, because we are using `depends_on` here, it will still wait for dependencies to start I guess. Not sure if there is much we can do in that respect, even if you try to point to a different Couchbase server host it may still download the image for the Couchbase server defined in the docker-compose file, even if it doesn't need to run it.

osfameron replied:
yeah. oddly it doesn't *stop* the db when you ctrl-c to stop it though.
I'll rewrite to use `docker`

osfameron committed:
Tweaks

osfameron committed:
Improve "mix-and-match" docs

osfameron committed:
Improve README / startup

 * Added nodemon to dev-dependencies.
 * Reformatted code blocks
 * Added up-to-date transcript
 * tweaks for clarity and narrative

maria-robobug commented on README.md:
```diff
@@ -11,179 +10,188 @@ for applicable air flight routes from a previously populated database. An additi
+We've integrated Swagger/OpenApi version 3 documentation which can be accessed on the backend at `http://localhost:8080/apidocs`
 
-It will also separate user account data such as flight bookings and credentials in the `tenant_agent_XX scopes`.
+More detail on the design of the API can be found  at [https://github.com/couchbaselabs/try-cb-frontend/blob/master/documentation/try-cb-api-spec-v2.adoc]
```
I don't think we should point to this as it is going to be old or deprecated. The swagger API should probably be enough.

osfameron replied:
Good call! Are there details in the doc that should be added to the Swagger?
e.g. unlike most API docs, which are purely reference, as this is an example, it's useful to give some description of what's happening behind the scenes like the doc does.
(though not actual code samples like that has)

maria-robobug replied:
Yeah we could add more to the descriptions in swagger for parameters etc, like so: https://github.com/couchbaselabs/try-cb-frontend/blob/master/documentation/try-cb-api-spec-v2.adoc#211-parameters.

Not sure if we need to show the queries, as you can see them in the code anyways, and it could change in future...

I do wonder if more information on the application might be best suited here though: https://docs.couchbase.com/nodejs-sdk/current/hello-world/sample-application.html

maria-robobug commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+      - db
+    ports:
+      - 8080:8080
+    environment:
+      - CB_HOST
+      - CB_USER
+      - CB_PSWD
+      - COLUMNS
```
Don't think we even need these env variables now, since we have `mix-and-match.yml` for running with different configurations.

osfameron replied:
ooo, good catch.

maria-robobug commented on README.md:
````diff
@@ -11,179 +10,188 @@ for applicable air flight routes from a previously populated database. An additi
-```
-git clone https://github.com/couchbaselabs/try-cb-nodejs.git
-```
+    ```
````
Looks like the indentation on the block doesn't work, probably it was because this was a in bullet points previously, so it worked then.

osfameron replied:
hmm, so we could use *just* indentation, or the triple-backtick github extension...
that said, https://github.github.com/gfm/#fenced-code-blocks does say 

> Fences can be indented. If the opening fence is indented, content lines will have equivalent opening indentation removed, if present

so not sure why it's not stripping the fencing

osfameron replied:
@simon-dew notes that fenced blocks can only be indented 0-3 characters. At 4, the "this is an indented block" behaviour kicks in.

osfameron committed:
Tidy README

Remove code-fencing from examples, as they're already indented.
(Though https://github.github.com/gfm/#fenced-code-blocks says

> Fences can be indented. If the opening fence is indented, content lines will
> have equivalent opening indentation removed, if present:

it seems not to strip them out)

Remove link to API docs in frontend repo, as we're now using the Swagger
as the canonical (and up-to-date) source.

osfameron committed:
Docker compose environment variables

We don't need the CB_* variables for the main docker-compose.yml,
as we expect any customizations to be done using mix-and-match.yml.

Also removed COLUMNS which was intended to make formatting of the
wait-for-couchbase output nicer, but seems not to be passed through,
so the default of 80 is used instead.)

maria-robobug commented:
Probably good idea to exclude `.out` files in the `.gitignore` to avoid committing them, we only want these for local use.

osfameron committed:
Exclude .out files

We don't want to commit e.g. hotel-index.out or wait-for-couchbase.out
as they are temporary work files.

osfameron commented:
> Probably good idea to exclude `.out` files in the `.gitignore` to avoid committing them, we only want these for local use.

Done. I did experiment with `mktemp` (which works fine, the Docker image doesn't have a /tmp so can't rely on that) for wait-for-couchbase, but then realised I should do it for the hotel-index too for consistency, and thought it seemed excessive.
(Thinking that might be wrong, mktemp feels like "the correct" thing to do, so could refactor if we want to.)

osfameron committed:
Update swagger.json
